File: binary_search/BinarySearchRecursive.java
import java.util.Arrays;

public class BinarySearchRecursive{
    public static int binary_search_recursive(int[] arr, int item, int left, int right){

        // Base case
        if (left > right){
            return -1;
        }

        int mid = left + (right - left) / 2;

        int guess = arr[mid];

        if (guess == item){
            return mid;
        }

        else if(item > guess){
            return binary_search_recursive(arr, item, mid + 1, right);
        }

        else {
            return binary_search_recursive(arr, item, left, mid - 1);
        }

    }

    public static void main(String[] args) {
        int array [] = {6, 4, 7, 2, 8, 1, 3};
        Arrays.sort(array);
        int left = 0;
        int right = array.length - 1;
        System.out.println(binary_search_recursive(array, 3, left, right));
        System.out.println(binary_search_recursive(array, 10, left, right));
    }
}

File: binary_search/binary_search_recursive.py
def binary_search_recursive(arr, item, left=None, right=None):
    if left is None:
        left = 0

    if right is None:
        right = len(arr) - 1

    # Base case
    if left > right:
        return None

    mid = left + (right - left) // 2
    #print(mid)

    guess = arr[mid]

    if guess == item:
        return mid

    elif item > guess:
        return binary_search_recursive(arr, item, mid + 1, right)

    else:
        return binary_search_recursive(arr, item, left, mid - 1)

list = [4, 5, 6, 7, 8, 3, 2, 9]
sort_list = sorted(list)
print(sort_list)
print(binary_search_recursive(sort_list, 7))
print(binary_search_recursive(sort_list, -1))
